### ISSUE TITLE
Defining resolvers creates circular dependency

### ISSUE THREAD

**Issue by JJK801** [aws-amplify/amplify-backend#1043]
### Environment information

```plain text
System:
  OS: macOS 13.3.1
  CPU: (10) arm64 Apple M2 Pro
  Memory: 121.36 MB / 16.00 GB
  Shell: /bin/zsh
Binaries:
  Node: 20.0.0 - ~/.nvm/versions/node/v20.0.0/bin/node
  Yarn: 1.22.19 - ~/.nvm/versions/node/v20.0.0/bin/yarn
  npm: 9.6.4 - ~/.nvm/versions/node/v20.0.0/bin/npm
  pnpm: undefined - undefined
NPM Packages:
  @aws-amplify/backend: Not Found
  @aws-amplify/backend-cli: Not Found
  aws-amplify: Not Found
  aws-cdk: Not Found
  aws-cdk-lib: Not Found
  typescript: Not Found
AWS environment variables:
  AWS_STS_REGIONAL_ENDPOINTS = regional
  AWS_NODEJS_CONNECTION_REUSE_ENABLED = 1
  AWS_SDK_LOAD_CONFIG = 1
No CDK environment variables
```

### Description

I'm trying to add a custom resolver, but i'm unable to do it because of circular dependencies.

This is probably because graphQL API and functions creation are done on different parallel stacks.

Here is a sample code:

```typescript
export const data = defineData({
  schema:  {
    Plan: a.customType({
        id: a.string().required(),
        // https://github.com/aws-amplify/amplify-category-api/issues/2126
        provider: a.string().required(), // .enum(['stripe'])
        active: a.boolean().required(),
        // https://github.com/aws-amplify/amplify-category-api/issues/2126
        period: a.string().required(), //.enum(['day', 'week', 'month', 'year']), 
        periodCount: a.integer().required(),
        priceCents: a.integer().required(),
        currency: a.string().required()
    }),
    listPlans: a.query()
        .returns(a.ref('Plan'))
        .authorization([
            a.allow.private()
        ])
        .function('listPlans')
  },
  functions: {
    listPlans: defineFunction({
      entry: '../functions/resolver/plan/handler.ts',
      environment: {
          STRIPE_SECRET_KEY: secret('STRIPE_SECRET_KEY')
      }
    })
  }
});
```

which results in:

```
❌ Deployment failed: Error [ValidationError]: Circular dependency between resources: [data7552DF31, function1351588B]
```

**Comment by edwardfoyle** [comment]
Hi @JJK801, do you have any other functions in your project that have a dependency on the AppSync API? (such as a function that has been granted access to the API). That would create the circular dependency you are seeing.

**Comment by JJK801** [comment]
hi @edwardfoyle

Yes, i made function some functions (and created a role allowing api access) to handle third party services events via event bridge . you're right that the problem should have to do with it.

i probably must set those functions out of the amplify context (preventing to use the nice `secret` integration 😫).

Do you see another option ?

About the data access, the `allowListedRoleNames` strategy to allow full access is really a pain, because we have to harcode a role name to do it (i'm actually bootstraping, so i've a single sandbox, but i see deployment errors coming when i will deploy new env)

**Comment by edwardfoyle** [comment]
Thanks for the follow up! We are actively working on a new access pattern to grant lambda access to other backend resources (including data) that will avoid this circular dependency issue. Once this lands, it will hopefully enable the use-cases you are describing without having to define the functions outside of `defineFunction`. The first iteration of this pattern has been implemented for storage [here](https://github.com/aws-amplify/amplify-backend/pull/969) and we will be expanding it to data as well.

**Comment by JJK801** [comment]
It looks great 👍

**Comment by JJK801** [comment]
@edwardfoyle now i've extracted my data interactors from amplify functions it's working without circular deps 👍 

Another issue i just had is that my custom model had no authorizations and i cannot set it on the whole type (as for models), so i had to authorize each field.

**Comment by edwardfoyle** [comment]
Can you open a new issue for that in the [data repo](https://github.com/aws-amplify/amplify-category-api/issues/new/choose)? They should be able to help with that.

**Comment by ideen1** [comment]
> Thanks for the follow up! We are actively working on a new access pattern to grant lambda access to other backend resources (including data) that will avoid this circular dependency issue. Once this lands, it will hopefully enable the use-cases you are describing without having to define the functions outside of `defineFunction`. The first iteration of this pattern has been implemented for storage [here](https://github.com/aws-amplify/amplify-backend/pull/969) and we will be expanding it to data as well.

Hey @edwardfoyle , this feature also helps me out a lot. I noticed it is present in the latest beta release and have been doing some testing with it. It seems to work and deploy as expected, however, I am not sure how I should pass the ARN for the bucket into the function so that it can be used. Obviously using backend.storage would not work since the lambda does not have access to that at runtime. Is there a solution to pass the ARN of the bucket to the lambda through an ENV variable or some other solution?

**Comment by JJK801** [comment]
@ideen1 you can do it in the `backend.ts` file.

```typescript
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';

// [...]

const backend = defineBackend(...);

const myFn = (backend.<pathToLambdaRessource>.lambda as NodejsFunction);
const myBucket = backend.<pathToStorageBucketResource>;

myFn.addEnvironment('BUCKET_ARN', myBucket.bucketArn)
```

**Comment by edwardfoyle** [comment]
Hi @ideen1, the name of the bucket is automatically provided to the function in an environment variable named `<storageName>_BUCKET_NAME`. We've recently updated our docs to call out this new behavior: https://docs.amplify.aws/gen2/build-a-backend/storage/#grant-function-access

**Comment by edwardfoyle** [comment]
Hi @JJK801, this is now available in the latest version of @aws-amplify/backend@beta. We've also added docs on how to set this up [here](https://docs.amplify.aws/gen2/build-a-backend/data/customize-authz/grant-lambda-function-access-to-api/)

Note that the experience for configuring the data client in the lambda is still a work in progress (which I see you have also discussed [here](https://github.com/aws-amplify/amplify-backend/issues/1056))
